Remove :bigquery driver

jeff303 opened the pull request:
Add data migration to convert existing Database instances from :bigquery to :bigquery-cloud-sdk (with error log if using outdated OAuth mechanisms)

Removing references to :bigquery from various places in the code

jeff303 committed:
Remove :bigquery driver

Add data migration to convert existing Database instances from :bigquery to :bigquery-cloud-sdk (with error log if using outdated OAuth mechanisms)

Removing references to :bigquery from various places in the code

jeff303 committed:
Remove from modules/drivers/deps.edn

jeff303 committed:
Fix data migration test

Fix unused namespace decl

jeff303 committed:
Remove stuff from CircleCI that is running the old driver

Remove `driver-switch-test` since there is no practical way to run this anymore (since we can't initialize the old driver)

jeff303 committed:
Fix namespace decl

camsaul commented on src/metabase/db/data_migrations.clj:
```diff
@@ -369,6 +369,19 @@
                                  [:like
                                   :dashcard.visualization_settings "%\"click_link_template\":%"]]})))
 
+(defmigration ^{:author "jeff303" :added "0.43.0"} remove-bigquery-driver
+  (doseq [{:keys [:details] :as bigquery-db} (db/select Database :engine "bigquery")]
+    (when-not (empty? (filter some? ((juxt :auth-code :client-id :client-secret) details)))
+      (log/errorf (str "Database ID %d used the :bigquery driver with OAuth style authentication scheme, which cannot"
+                       " be automatically migrated to :bigquery-cloud-sdk; leaving authentication options nil in the"
+                       " db details, which must be corrected by an admin (by setting a service-account-json), so that"
+                       " sync and query can work again")
+                  (u/the-id bigquery-db)))
+    (let [migrated-details (-> (assoc details :engine :bigquery-cloud-sdk) ; switch to new driver
+                               ;; remove all OAuth related keys
+                               (dissoc :auth-code :client-id :client-secret :access-token :refresh-token))]
+      (db/update! Database (u/the-id bigquery-db) :details migrated-details))))
```
I really really would prefer this be done in a Liquibase migration. Are you 100% sure we can't just migrate the driver in Liqubase e.g.

```sql
UPDATE metabase_database
SET engine = 'bigquery-cloud-sdk'
WHERE engine = 'bigquery'
```

And then detect and correct this situation (legacy OAuth details) in the `normalize-db-details` method?

camsaul replied:
Actually, wouldn't it be sufficient just to ignore those keys entirely? Do you even need to remove them at all?

jeff303 replied:
Yeah, this is feasible.  I originally got stuck into thinking I could register `normalize-db-details` for `:bigquery`, but that obviously wouldn't work if the driver (which needs initialization) is being removed.

camsaul commented on src/metabase/db/data_migrations.clj:
```diff
@@ -369,6 +369,19 @@
+                       " db details, which must be corrected by an admin (by setting a service-account-json), so that"
+                       " sync and query can work again")
+                  (u/the-id bigquery-db)))
+    (let [migrated-details (-> (assoc details :engine :bigquery-cloud-sdk) ; switch to new driver
```
`engine` is a separate top-level column in `Database`. It's not inside `details`

jeff303 replied:
Oops, you'd think I would remember this by now.

camsaul reviewed (changes requested):
It looks like the data migration doesn't actually update `Database.engine`. Either way, that part should 100% be done as a Liquibase migration. I don't think we **need** to remove the old OAuth keys from details but if we do we should do it in `normalize-details` rather than in a Clojure-land data migration.

I would vote for just ignoring them, that way if someone switches branches and downgrades they'd just have to switch the driver back and wouldn't have to re-enter all of their OAuth credentials 

Other than that everything else is looking good

camsaul commented on src/metabase/db/data_migrations.clj:
```diff
@@ -180,7 +180,7 @@
 (defmigration ^{:author "camsaul", :added "0.30.0"} add-legacy-sql-directive-to-bigquery-sql-cards
   ;; For each BigQuery database...
-  (doseq [database-id (db/select-ids Database :engine "bigquery")]
+  (doseq [database-id (db/select-ids Database :engine "bigquery")] ; TODO: keep this?
```
We should go ahead and take this out now. Since we'd be upgrading to the new driver before we ever reach this point this migration will effectively always be a no-op. 

If anyone migrates from 29 or older to 43 or newer they'll have to manually add `#legacy` directives to their BigQuery SQL queries written in legacy SQL

jeff303 replied:
Done.

jeff303 committed:
Merge branch 'master' into remove-legacy-bigquery-driver

jeff303 committed:
Switch from data to schema (Liquibase) migration, and move test accordingly

Move logging of the OAuth error to the `driver/normalize-db-details` impl for `:bigquery-cloud-sdk`

jeff303 committed:
Remove now defunct data migration as per PR feedback

jeff303 committed:
Remove unused require

camsaul reviewed (approved):
Awesome. This LGTM. Thanks for doing this

jeff303 committed:
Merge branch 'master' into remove-legacy-bigquery-driver

jeff303 committed:
Merge branch 'master' into remove-legacy-bigquery-driver

camsaul committed:
Fix the mega oof

camsaul commented:
@jeff303 Done
